lexer/token.go

```diff
@@ -0,0 +1,209 @@
+package lexer
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+)
+
+// TokenType custom type defines the Token type attribute.
+type TokenType int
+
+const (
+	// TokenTypeLiteral is the type value assigned to literal tokens
+	TokenTypeLiteral TokenType = iota
+	// TokenTypeGroup is the type value assigned to group tokens
+	TokenTypeGroup
+)
+
+// Token struct envolves any type of token (literal or group).
+type Token struct {
+	Type    TokenType
+	Literal string
+	Childs  *Group
+}
+
+// MarshalJSON method helps to encode the current token into a JSON with just
+// the important information, discarding internal information.
+func (t *Token) MarshalJSON() ([]byte, error) {
+	if t.IsLiteral() {
+		// if it is a literal token include just the token symbol
+		return json.Marshal(map[string]any{"literal": t.Literal})
+	}
+	// if it is a group token, include just the child tokens
+	return json.Marshal(map[string]any{"childs": t.Childs})
+}
+
+// IsLiteral method returns if the current token is a literal token (or a group
+// token).
+func (t *Token) IsLiteral() bool {
+	return t.Type == TokenTypeLiteral
+}
+
+// String function returns the string version of the current token. It can be
+// used as predicated input to generate the same token.
+func (t *Token) String() string {
+	// if it is literal envolves between literals delimiters
+	if t.IsLiteral() && t.Literal != "" {
+		return fmt.Sprintf("'%s'", t.Literal)
+	}
+	// if it is group and it has not childs, return empty
+	if t.Childs == nil || len(t.Childs.Tokens) < 2 {
+		return ""
+	}
+	// return string group
+	return fmt.Sprintf("(%s %s %s)",
+		t.Childs.Tokens[t.Childs.firstToken],
+		t.Childs.Operator,
+		t.Childs.Tokens[t.Childs.secondToken],
+	)
+}
+
+// Equals method return if the a token is equal to b token. It compares the
+// between both tokens string versions.
+func (a *Token) Equals(b *Token) bool {
+	return a.String() == b.String()
+}
+
+// AllGroups method returns all the group tokens from the current token to the
+// last child, including it. The method calls itself recursively if the current
+// token is a group token.
+func (t *Token) AllGroups() []*Group {
+	if t.IsLiteral() || t.Childs == nil {
+		return nil
+	}
+	res := []*Group{t.Childs}
+	for _, t := range t.Childs.Tokens {
+		if !t.IsLiteral() {
+			res = append(res, t.AllGroups()...)
+		}
+	}
+	return res
+}
+
+// AllLiterals method returns all the literal tokens from the current token to
+// the last child, including it. The method calls itself recursively if the
+// current token is a group token.
+func (t *Token) AllLiterals() []string {
+	if t.IsLiteral() {
+		return []string{t.Literal}
+	}
+	if t.Childs == nil {
+		return nil
+	}
+	res := []string{}
+	commit := func(i string) {
+		exists := false
+		for _, r := range res {
+			if r == i {
+				exists = true
+				break
+			}
+		}
```

Comment:
I don't know if the performance really matters here because we don't expect too long or deep predicates, or not enough complex to care about it, but it is an interest improvement.

I was thinking about implement the list that you suggested but I have a question. As I designed the predicates, they are represented by a `Token` struct, that could contain other `Token` structs, creating a *tree* of `Token`s. Must every `Token` have that sorted list instead instead to loop throw the entire tree to get every token?

Reply:
I will open a issue with this to keep it in mind in the future.